File: UI/src/main/java/ca/warp7/rt/feature/scanner/ScannerFeature.java
package ca.warp7.rt.feature.scanner;

import ca.warp7.rt.core.feature.Feature;
import ca.warp7.rt.core.feature.FeatureItemTab;
import ca.warp7.rt.core.feature.FeatureTabFactory;
import ca.warp7.rt.core.feature.FeatureUtils;
import javafx.collections.FXCollections;
import javafx.scene.Parent;

import java.util.List;

import static ca.warp7.rt.core.feature.FeatureItemTab.Group;

public class ScannerFeature implements Feature {

    private FeatureTabFactory factory = new FeatureTabFactory("fas-camera:16:gray", getFeatureId(), Group.SingleTab);

    private ScannerController controller;

    private void setController(ScannerController controller) {
        this.controller = controller;
    }

    @Override
    public List<FeatureItemTab> getLoadedTabs() {
        return FXCollections.singletonObservableList(factory.get("QR Scanner", ""));
    }

    @Override
    public String getFeatureId() {
        return "scanner";
    }

    @Override
    public Parent onOpenTab(FeatureItemTab tab) {
        return FeatureUtils.loadParent("/ca/warp7/rt/feature/scanner/Scanner.fxml", this::setController);
    }

    @Override
    public boolean onCloseRequest() {
        controller.stopCameraStream();
        return true;
    }
}

File: UI/src/main/kotlin/ca/warp7/rt/feature/scanner/ScannerFeature.kt
package ca.warp7.rt.feature.scanner

import ca.warp7.rt.core.feature.Feature
import ca.warp7.rt.core.feature.FeatureItemTab
import ca.warp7.rt.core.feature.FeatureItemTab.Group
import ca.warp7.rt.core.feature.FeatureTabFactory
import ca.warp7.rt.core.feature.FeatureUtils
import javafx.collections.FXCollections
import javafx.scene.Parent

class ScannerFeature : Feature {

    private val factory = FeatureTabFactory("fas-camera:16:gray", featureId, Group.SingleTab)

    private var controller: ScannerController? = null

    private fun setController(controller: ScannerController) {
        this.controller = controller
    }

    override fun getLoadedTabs(): List<FeatureItemTab> {
        return FXCollections.singletonObservableList(factory.get("QR Scanner", ""))
    }

    override fun getFeatureId(): String {
        return "scanner"
    }

    override fun onOpenTab(tab: FeatureItemTab): Parent? {
        return FeatureUtils.loadParent<ScannerController>("/ca/warp7/rt/feature/scanner/Scanner.fxml") {
            this.setController(it)
        }
    }

    override fun onCloseRequest(): Boolean {
        controller!!.stopCameraStream()
        return true
    }
}
